New resource r/aws_wafv2_web_acl_association

pvanbuijtene opened the pull request:
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #11177
Relates #11046

This PR can not be merged before  #12688 because of copied dependencies.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
```release-note
* **New Resource**: `aws_wafv2_web_acl_association` (#12698)
```

Output from acceptance testing:

```
make testacc TEST=./aws TESTARGS='-run=TestAccAwsWafv2WebACLAssociation'
...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAwsWafv2WebACLAssociation -timeout 120m
=== RUN   TestAccAwsWafv2WebACLAssociation_Basic
=== PAUSE TestAccAwsWafv2WebACLAssociation_Basic
=== CONT  TestAccAwsWafv2WebACLAssociation_Basic
--- PASS: TestAccAwsWafv2WebACLAssociation_Basic (83.75s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       84.276s
```

anGie44 commented on aws/resource_aws_wafv2_web_acl_association.go:
```diff
@@ -0,0 +1,210 @@
+	if r.Service == "apigateway" {
+		return aws.String(wafv2.ResourceTypeApiGateway)
+	} else if r.Service == "elasticloadbalancing" {
+		return aws.String(wafv2.ResourceTypeApiGateway)
```
*`wafv2.ResourceTypeApplicationLoadBalancer`

pvanbuijtene replied:
Ouch

anGie44 commented on aws/resource_aws_wafv2_web_acl_association.go:
```diff
@@ -0,0 +1,210 @@
+		return fmt.Errorf("error reading resource ID: %s", err)
+	}
+
+	params := &wafv2.ListResourcesForWebACLInput{
```
I believe we could also help reduce the search here by using the`wafv2.GetWebACLForResource`  method with the knwon `resourceArn`

anGie44 commented on aws/resource_aws_wafv2_web_acl_association.go:
```diff
@@ -0,0 +1,210 @@
+
+func resourceAwsWafv2WebACLAssociationRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).wafv2conn
+	webAclArn, resourceArn, err := resourceAwsWafv2ACLAssociationDecodeId(d.Id())
```
I believe we can read directly from the resource data here since in `Create` -> `Read`  the values are set and at `Import` time, the logic w/in the Importer's `State` func will store the values for us

pvanbuijtene commented:
@anGie44 the changes have been applied. I removed the WaitForState code because the test runs I did all succeeded without it. I also removed the the retry on internal error as this will be retried by the retry handling from the other PR.

DrFaust92 commented on aws/resource_aws_wafv2_web_acl_association_test.go:
```diff
@@ -0,0 +1,142 @@
+			{
+				Config: testAccAwsWafv2WebACLAssociationConfig(testName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAWSWafv2WebACLAssociationExists(resourceName),
```
add tests for arn using the test func, i.e. `testAccMatchResourceAttrRegionalARN()`

pvanbuijtene commented:
@DrFaust92 the changes are applied.

Output from acceptance tests:
```
--- PASS: TestAccAwsWafv2WebACLAssociation_Basic (108.71s)
--- PASS: TestAccAwsWafv2WebACLAssociation_Disappears (138.48s)
```

anGie44 commented on aws/resource_aws_wafv2_web_acl_association.go:
```diff
@@ -0,0 +1,133 @@
+		WebACLArn:   aws.String(webAclArn),
+	}
+
+	err := resource.Retry(2*time.Minute, func() *resource.RetryError {
```
we could also emulate the pattern in the `wafv2_regex_pattern_set` PR of adding to the err handlers in `config.go` for this error type/operation combination

pvanbuijtene replied:
@anGie44 are you sure you want to move this?
This is not really a generic error but more related to this resource so I'm not sure if we should move this, also according to what @bflad wrote here [here](https://github.com/terraform-providers/terraform-provider-aws/pull/12119#discussion_r419806963).
If you still want it to be moved,let me know and I'll move it to `config.go`

anGie44 replied:
ah sorry I glazed over the exception type here - yep you're totally right, it should stay here as it's specific to the resource

anGie44 commented on aws/resource_aws_wafv2_web_acl_association_test.go:
```diff
@@ -0,0 +1,166 @@
+		})
+
+		// Return nil if the Web ACL Association is already destroyed
+		if awsErr, ok := err.(awserr.Error); ok {
```
we can use the` isAWSErr()` helper function here instead to help with the error code comparison

anGie44 commented:
hi @pvanbuijtene! thanks again for your changes. I've just added some final comments building off what's been reviewed in other wafv2 PRs when it comes to error handling and a recommendation to use the AWS Go SDK's built-in error helper.

anGie44 commented on aws/resource_aws_wafv2_web_acl_association_test.go:
```diff
@@ -0,0 +1,163 @@
+		if isAWSErr(err, wafv2.ErrCodeWAFNonexistentItemException, "") {
+			return nil
+		}
+
```
we should also check here if `err != nil` and return it if so after the initial `isAwsErr` check

anGie44 commented on aws/resource_aws_wafv2_web_acl_association_test.go:
```diff
@@ -0,0 +1,171 @@
+		if err == nil {
+			id := fmt.Sprintf("%s,%s", aws.StringValue(resp.WebACL.ARN), rs.Primary.Attributes["resource_arn"])
+			if id == rs.Primary.ID {
+				return fmt.Errorf("WAFv2 IPSet %s still exists", rs.Primary.ID)
```
small typo here - should be WebACLAssociation, other wise looks good 👍

pvanbuijtene commented:
@anGie44 the code is removed.

anGie44 commented on website/docs/r/wafv2_web_acl_association.html.markdown:
```diff
@@ -0,0 +1,87 @@
+
+## Import
+
+WAFv2 Web ACL Association can be imported using the Web ACL ARN, and the resource ARN e.g.
```
Nit: let's format this import sytnax with: 
```suggestion
WAFv2 Web ACL Association can be imported using `WEB_ACL_ARN,RESOURCE_ARN` e.g.
```

anGie44 commented on website/docs/r/wafv2_web_acl_association.html.markdown:
```diff
@@ -0,0 +1,87 @@
+
+# Resource: aws_wafv2_web_acl_association
+
+Creates an WAFv2 Web ACL Association.
```
Ah i missed this article 😅 
```suggestion
Creates a WAFv2 Web ACL Association.
```

pvanbuijtene replied:
... no worries ⏳

anGie44 commented:
Thanks again for your work @pvanbuijtene! 

Note: holding off acceptance testing as continent on #11176

bflad commented:
Waiting to review until #12688 is merged.

anGie44 commented on aws/resource_aws_wafv2_web_acl_association_test.go:
```diff
@@ -0,0 +1,175 @@
+					testAccCheckResourceDisappears(testAccProvider, resourceAwsWafv2WebACLAssociation(), resourceName),
+				),
+				ExpectNonEmptyPlan: true,
+			},
```
we could additionally verify the case where the resource part of the webacl association disappears e.g.
```
{
       Config: testAccAwsWafv2WebACLAssociationConfig(testName),
       Check: resource.ComposeTestCheckFunc(
              testAccCheckAWSWafv2WebACLAssociationExists(resourceName),
	      testAccCheckResourceDisappears(testAccProvider, resourceAwsApiGatewayStage(), "aws_api_gateway_stage.test"),
       ),
       ExpectNonEmptyPlan: true,
},
```

anGie44 commented:
hi @pvanbuijtene, please also rebase when you have a chance. while the checks here in Github appear to pass, when I pulled down your branch locally the linter still catches the extra param error in the `web_acl_test` file

anGie44 commented:
Awesome, looks great @pvanbuijtene 🚀 
Output of acceptance tests:
```
--- PASS: TestAccAwsWafv2WebACLAssociation_Basic (85.59s)
--- PASS: TestAccAwsWafv2WebACLAssociation_Disappears (115.47s)
```

jithinjose2004 commented:
Hey @anGie44  ,i noticed that the aws_wafv2_web_acl_association does not support CloudFront Distribution association . Is it something planned for future or totally ignoring CF ?

anGie44 commented:
Hi @jithinjose2004! For CloudFront Distribution associations, this can be done within your `aws_cloudfront_distribution` resource if also provisioned by Terraform. AWS docs atm don't show support for this within WebACL resources (https://docs.aws.amazon.com/waf/latest/APIReference/API_AssociateWebACL.html) but of course this could change in the future. One thing to note if configuring a `aws_cloudfront_distribution` to associate a WAFv2 WebACL is to make sure to pass the WebACL's ARN (not ID) to the `web_acl_id` field as this naming convention originally applied only to WAF Classic.  

Hope this helps! If further questions arise, let me know 😃

jithinjose2004 commented:
Hi @anGie44 , Thank you for your reply. right now web_acl_arn only supports api_gateway_arn or alb_arn. I wanted to associate cloudfront_ditribution_arn. Hope terraform will support this in future. 

Another issue is , If I make changes/add new rule to the waf2 acl script , terraform recreates the whole waf_acl and  this will remove the existing distribution association done manually. This makes it hard :) .  Am i doing it in the wrong way ?
